### PR TITLE
14 Refine bayesian ridge implementation

### PR DESCRIPTION
**Pull request by martinakaduc**
I finished cleaning up the implementation of bayesian ridge. Now, it is less than 100 lines of code.

### CODE REVIEW

**Comment by martinakaduc** [comment]
> @martinakaduc @LMorlok This test tests the optimizer (i.e., the loss is low) instead of testing whether the core functions (forward, posterior, get_fantasy_model) are implemented correctly. Hence, this test cannot determine whether the class BayesRidgeModel is implemented correctly. I imagine someone could write a wrong forward function while still having a low loss. It would help if you designed better tests. Beyond that, plotting doesn't help with (automated) testing here either.

Currently I don't have any ideas for testing correctness. Can you suggest some ideas?

**Comment by sangttruong** [comment]
You can test the correctness as follows:
Instantiate a random weight and a random input. Compute the forward pass by hand (it can be done in close form). Do an assert to see if the computational output from forward matches your result. Do the same thing for the posterior. Does this make sense?

**Comment by martinakaduc** [comment]
> You can test the correctness as follows: Instantiate a random weight and a random input. Compute the forward pass by hand (it can be done in close form). Do an assert to see if the computational output from forward matches your result. Do the same thing for the posterior. Does this make sense?

Please do have a review

**Comment by sangttruong** [comment]
Can you explain what you did for the test here? 

```
        model.br_model.coef_ = rand_coef_
        model.br_model.intercept_ = rand_intercept_
        model.br_model.sigma_ = rand_sigma_
        model.br_model.alpha_ = rand_alpha_

        x_test = np.random.rand(100, dim)
        y_test_dist = model.forward(torch.tensor(x_test))
        y_test_mean = y_test_dist.mean
        y_test_std = y_test_dist.covariance_matrix.diag()

        y_test_mean_truth = x_test @ rand_coef_ + rand_intercept_
        sigmas_squared_data = (np.dot(x_test, rand_sigma_) * x_test).sum(axis=1)
        y_test_std_truth = np.sqrt(sigmas_squared_data + (1.0 / rand_alpha_))

```

**Comment by sangttruong** [comment]
@martinakaduc Can you explain what you did for the test here? 
```
        model.br_model.coef_ = rand_coef_
        model.br_model.intercept_ = rand_intercept_
        model.br_model.sigma_ = rand_sigma_
        model.br_model.alpha_ = rand_alpha_

        x_test = np.random.rand(100, dim)
        y_test_dist = model.forward(torch.tensor(x_test))
        y_test_mean = y_test_dist.mean
        y_test_std = y_test_dist.covariance_matrix.diag()

        y_test_mean_truth = x_test @ rand_coef_ + rand_intercept_
        sigmas_squared_data = (np.dot(x_test, rand_sigma_) * x_test).sum(axis=1)
        y_test_std_truth = np.sqrt(sigmas_squared_data + (1.0 / rand_alpha_))
```

**Comment by martinakaduc** [comment]
> @martinakaduc Can you explain what you did for the test here?
> 
> ```
>         model.br_model.coef_ = rand_coef_
>         model.br_model.intercept_ = rand_intercept_
>         model.br_model.sigma_ = rand_sigma_
>         model.br_model.alpha_ = rand_alpha_
> 
>         x_test = np.random.rand(100, dim)
>         y_test_dist = model.forward(torch.tensor(x_test))
>         y_test_mean = y_test_dist.mean
>         y_test_std = y_test_dist.covariance_matrix.diag()
> 
>         y_test_mean_truth = x_test @ rand_coef_ + rand_intercept_
>         sigmas_squared_data = (np.dot(x_test, rand_sigma_) * x_test).sum(axis=1)
>         y_test_std_truth = np.sqrt(sigmas_squared_data + (1.0 / rand_alpha_))
> ```

This code block is used for testing `forward` function. `y_test_mean` and `y_test_std` are extracted from results of `forward` function.

I compute the `y_test_mean_truth` and `y_test_std_truth` manually using predefined coefficients. After that, I compare the results of forward function and the manually computed results to see whether they are exactly the same or not.
